### PR TITLE
Cpu main thread

### PR DESCRIPTION
**Pull request by rafaelfolco**

### CODE REVIEW

File: oslat-client [hunk 1]
```diff
@@ -104,8 +104,13 @@ echo -e "${CMD_OUTPUT}"
...
 
-# get the first CPU from HK_CPUS to use as the oslat main thread
-cpu_main_thread=$(echo ${HK_CPUS} | awk -F, '{ print $1 }')
+# get the first CPU from WORKLOAD_CPUS to use as the oslat main thread
+cpu_main_thread=$(echo ${WORKLOAD_CPUS} | awk -F, '{ print $1 }')
+echo "cpu-main-thread: ${cpu_main_thread}"
+
+# LATENCY_CPUS are the WORKLOAD_CPUS less the cpu main thread
+LATENCY_CPUS=$(echo ${WORKLOAD_CPUS} | cut -d ',' -f2-)
+echo "LATENCY_CPUS: ${LATENCY_CPUS}"
```
Comment (k-rister) [inline oslat-client:113]: I think we need to decide if we want to make this mandatory or optional (ie. controlled by a parameter).

Comment (k-rister) [reply oslat-client:113]: @rafaelfolco @atheurer ??

Comment (k-rister) [reply oslat-client:113]: I'm not sure what we want to do, I was hoping for some input from @atheurer ...

File: oslat-server-start [hunk 2]
```diff
@@ -66,14 +66,16 @@ echo "TEST_TILES: ${TEST_TILES}"
 
 CMD="taskset -c ${CPUS}"
 CMD+=" stress-ng"
+CMD+=" --taskset ${CPUS}"
```
Comment (k-rister) [inline oslat-server-start:69]: hmm, what is the different between the `taskset -c` and this?

File: oslat-client [hunk 3]
```diff
@@ -130,7 +150,7 @@ if [ $? != 0 ]; then
     exit_error "${output}"
 fi
 
-cmd="taskset -c ${HK_CPUS},${WORKLOAD_CPUS} oslat --duration ${duration} ${rtprio_opt} --cpu-list ${WORKLOAD_CPUS} --cpu-main-thread ${cpu_main_thread} ${trace_args}"
+cmd="taskset -c ${WORKLOAD_CPUS} oslat --duration ${duration} ${rtprio_opt} --cpu-list ${LATENCY_CPUS} --cpu-main-thread ${cpu_main_thread} ${trace_args}"
```
Comment (k-rister) [inline oslat-client:153]: Does `--cpu-main-thread` accept more than one CPU?  That could happen in the default case...

File: oslat-server-start (hunk 2 above)
Comment (rafaelfolco) [reply oslat-server-start:69]: none, same thing

File: oslat-client (hunk 3 above)
Comment (rafaelfolco) [reply oslat-client:153]: we can fix this by adding a new param "cpu-main-thread". Will keep the default as the hk cpus then.

**Review by k-rister — CHANGES_REQUESTED** [review]
Seems find except for some really weird indentation stuff -- mostly (only?) in the arg processing part.  Can you fix that?

**Review by k-rister — DISMISSED** [review]

**Comment by rafaelfolco** [comment]
I was able to reproduce the issue we're seeing in CI.
The 4th sample fails on client-start-end step (The first 3 samples worked just fine).
investigating...




`uuid: 1:9f15cab7-c0c8-4b76-96c2-7e2e14c024a6:1-4-1:client-start-end
going to run this roadblock command:  /usr/local/bin/roadblocker.py --role=follower --redis-server=10.26.9.242 --leader-id=controller --timeout=190 --redis-password=flubber --follower-id=client-1 --message-log=/tmp/tmp.uEGvulyupC/roadbl
ock-msgs/1-4-1:client-start-end.json --uuid=1:9f15cab7-c0c8-4b76-96c2-7e2e14c024a6:1-4-1:client-start-end
roadblock output BEGIN
[2024-12-05 19:33:38,328][    INFO] Current Time: 2024-12-05 at 19:33:38 UTC
[2024-12-05 19:33:38,328][    INFO] Wait-For: False
[2024-12-05 19:33:38,339][    INFO] Timeout: 2024-12-05 at 19:36:48 UTC
[2024-12-05 19:33:38,374][    INFO] Roadblock UUID: 1:9f15cab7-c0c8-4b76-96c2-7e2e14c024a6:1-4-1:client-start-end
[2024-12-05 19:33:38,374][    INFO] Role: follower
[2024-12-05 19:33:38,374][    INFO] Follower ID: client-1
[2024-12-05 19:33:38,374][    INFO] Leader ID: controller
[2024-12-05 19:33:38,374][    INFO] Abort: False
[2024-12-05 19:33:38,394][    INFO] Initiator: False
[2024-12-05 19:33:38,394][    INFO] Waiting for roadblock initialization to complete
[2024-12-05 19:33:38,406][    INFO] Roadblock is initialized
[2024-12-05 19:33:38,406][    INFO] Creating personal bus
[2024-12-05 19:33:38,424][    INFO] Sending 'follower-online' message
[2024-12-05 19:33:38,441][    INFO] Received 'global-bus-created' message
[2024-12-05 19:33:38,441][    INFO] Received 'timeout-ts' message
[2024-12-05 19:33:38,441][    INFO] The new timeout value is in 148 seconds
[2024-12-05 19:33:38,442][    INFO] Timeout: 2024-12-05 at 19:36:06 UTC
[2024-12-05 19:33:38,442][    INFO] Received 'followers-bus-created' message
[2024-12-05 19:33:38,442][    INFO] Received 'personal-bus-created' message
[2024-12-05 19:36:06,461][CRITICAL] Roadblock failed with timeout
[2024-12-05 19:36:06,465][    INFO] Cleaning up
[2024-12-05 19:36:06,465][    INFO] Closing connection pool
[2024-12-05 19:36:06,489][    INFO] Exiting
[2024-12-05 19:36:06,489][    INFO] Roadblock Completed with an Error
roadblock output END
roadblock exit code: 3`

File: oslat-client [hunk 4]
```diff
@@ -91,7 +102,14 @@ echo "about to run: ${cmd}"
...
 echo -e "${CMD_OUTPUT}"
 WORKLOAD_CPUS=$(echo -e "${CMD_OUTPUT}" | grep "final cpus:" | awk '{ print $3 }')
-echo "WORLOAD_CPUS: ${WORKLOAD_CPUS}"
+echo "WORKLOAD_CPUS: ${WORKLOAD_CPUS}"
+
+if [[ -n "${cpu_list}" ]]; then
+    echo "Overriding WORKLOAD_CPUS: $WORKLOAD_CPUS w/ cpu-list: ${cpu_list}"
+    WORKLOAD_CPUS=${cpu_list}
+    echo "WORKLOAD_CPUS = $WORKLOAD_CPUS"
+fi
```
Comment (k-rister) [inline oslat-client:111]: So I'm a little curious what the use case is for this option?

Comment (k-rister) [reply oslat-client:111]: Right...but are we ever actually doing that?

Comment (rafaelfolco) [reply oslat-client:111]: No...
We always automatically guessed/detected the workload cpus and housekeeping cpus.  We currently do NOT allow the user to select a custom range of cpus. This is the [original code](https://github.com/perftool-incubator/bench-oslat/blob/master/oslat-client#L133):
`
cmd="taskset -c ${HK_CPUS},${WORKLOAD_CPUS} oslat --duration ${duration} ${rtprio_opt} --cpu-list ${WORKLOAD_CPUS} --cpu-main-thread ${cpu_main_thread} ${trace_args}"
`

We used to measure latency with the workload cpus. Then, more recently, we added the cpu-main-thread [param](https://github.com/perftool-incubator/bench-oslat/blob/master/oslat-client#L107C1-L108C60)
 by automatically selecting the first cpu from the workload cpus range:
`
# get the first CPU from HK_CPUS to use as the oslat main thread
cpu_main_thread=$(echo ${HK_CPUS} | awk -F, '{ print $1 }')
`
Now the proposal in this PR is to let the user specify in the JSON run file what is the cpu-list for measuring latency (--cpu-list param), and what is the houskeeping cpu range (--house-keeping param). Default behavior, if not specified, continue to be: automatically detect the workload/hk cpus.

Future PR to validate if the user range is valid.... For now we assume that if cpus are specified, user knows what he is doing.

**Review by k-rister — APPROVED** [review]

**Pull request merged**